File: lintcode/golang/1357_path_sum_II.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: a binary tree
 * @param sum: the sum
 * @return: the scheme
 */
func pathSum (root *TreeNode, sum int) [][]int {
    // Write your code here.
    if root == nil {
        return [][]int{}
    }
    
    var ret [][]int
    var dfs func(*TreeNode, int, []int)
    
    dfs = func(node *TreeNode, target int, path []int) {
        if node == nil {
            return
        }
        if node.Left == nil && node.Right == nil && target == node.Val {
            path = append(path, node.Val)
            ret = append(ret, append([]int{}, path...))
            return
        }
        
        path = append(path, node.Val)
        dfs(node.Left, target-node.Val, path)
        dfs(node.Right, target-node.Val, path)
        path = path[:len(path)-1]
    }
    
    dfs(root, sum, []int{})
    
    return ret
}

File: lintcode/python/1357_path_sum_II.py
#!/usr/bin/python -t

# dfs

# Time complexity: O(N)
# Space complexity: O(N)

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: a binary tree
    @param sum: the sum
    @return: the scheme
    """
    def pathSum(self, root, target):
        # Write your code here.
        result = []

        self.dfs(root, target, [], result)

        return result

    def dfs(self, node, target, permutation, result):
        if not node:
            return

        if target < 0:
            return

        if not node.left and not node.right and target == node.val:
            result.append(list(permutation + [node.val]))
            return

        permutation.append(node.val)
        self.dfs(node.left, target - node.val, permutation, result)
        self.dfs(node.right, target - node.val, permutation, result)
        permutation.pop()
